nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -0,0 +1,63 @@
+package nsis.file;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import generic.continues.GenericFactory;
+import ghidra.app.util.bin.ByteProvider;
+import ghidra.app.util.bin.format.FactoryBundledWithBinaryReader;
+import ghidra.app.util.bin.format.pe.PortableExecutable.SectionLayout;
+
+public class NsisExecutable {
+
+	public static final String NAME = "NULLSOFT_SCRIPTABLE_INSTALLER_SYSTEM";
+	
+	private FactoryBundledWithBinaryReader reader;
+	private long header_offset;
+	
+	public NsisExecutable() {}
+	
+	public static NsisExecutable createNsisExecutable(GenericFactory factory, ByteProvider bp, SectionLayout layout) {
+		NsisExecutable nsisExecutable = (NsisExecutable) factory.create(NsisExecutable.class);
+		nsisExecutable.initNsisExecutable(factory, bp, layout);
+		return nsisExecutable;
+	}
+	
+	private void initNsisExecutable(GenericFactory factory, ByteProvider bp, SectionLayout layout) {
+		this.reader = new FactoryBundledWithBinaryReader(factory, bp, true);
+		this.header_offset = -1;
+		findHeaderOffset();
+		//TODO Init sections and headers
+	}
+	
+	private void findHeaderOffset() {
+		if(this.header_offset == -1) {
+			for(long offset = 0; offset + NsisConstants.NSIS_MAGIC.length <= this.getFileLength(); offset++) {
+				try {
+					byte[] content = reader.readByteArray(offset, NsisConstants.NSIS_MAGIC.length);
+					if(Arrays.equals(NsisConstants.NSIS_MAGIC, content)) {
+						this.header_offset = offset;
+						return;
+					}
+				} catch (IOException e) {
+					return;
+				}
+			}
+		}
+	}
+	
+	public long getHeaderOffset() {
+		return this.header_offset;
+	}
+	
+	public long getFileLength() {
```

Comment:
I don't think this function is necessary. I think we can just call `reader.length()` directly. I don't think we want to hide the exception by returning 0. If we can't get the size of the file it points to a fundamental problem and it would be best to throw the exception.

Reply:
I think we will need to get the length of the NSIS file in the NSISLoader.java class, and since the loader is a private attribute of NSISexecutable I was thinking I would have to either create a function that returns the loader attribute or the length of the file. Do you think it would be better to return the loader attribute instead?

Reply:
Ah woups, it's true that the byteProvider is a parameter of NsisLoader.java critical functions, so we don't need to provide a getLength for NsisExecutable.java at least for now. Thanks for spotting that!